freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UsersStateDifference.java

```diff
@@ -113,6 +125,39 @@ public static UsersStateDifference forDeletedUser(String deletedUser, Collection
         return usersToAdd;
     }
 
+    public static ImmutableSet<String> calculateUsersWithCredentialsToUpdate(UmsUsersState umsState, UsersState ipaState,
+            boolean credentialsUpdateOptimizationEnabled) {
+        ImmutableSet<String> usersWithCredentialsToUpdate;
+
+        if (!credentialsUpdateOptimizationEnabled) {
+            usersWithCredentialsToUpdate = ImmutableSet.copyOf(umsState.getUsersState().getUsers().stream()
+                    .map(FmsUser::getName)
+                    .filter(username -> !FreeIpaChecks.IPA_PROTECTED_USERS.contains(username))
+                    .collect(Collectors.toSet()));
+        } else {
+            ImmutableSet.Builder<String> usersWithCredentialsToUpdateBuilder = ImmutableSet.builder();
+            umsState.getUsersState().getUsers().stream()
+                    .map(FmsUser::getName)
+                    .filter(username -> !FreeIpaChecks.IPA_PROTECTED_USERS.contains(username))
+                    .forEach(username -> {
+                        UserMetadata ipaUserMetadata = ipaState.getUserMetadataMap().get(username);
+                        if (ipaUserMetadata == null) {
+                            usersWithCredentialsToUpdateBuilder.add(username);
+                        } else {
+                            WorkloadCredential umsCredential = umsState.getUsersWorkloadCredentialMap().get(username);
+                            if (ipaUserMetadata.getWorkloadCredentialsVersion() < umsCredential.getVersion()) {
+                                usersWithCredentialsToUpdateBuilder.add(username);
+                            }
+                        }
+                    });
+            usersWithCredentialsToUpdate = usersWithCredentialsToUpdateBuilder.build();
+        }
+
+        LOGGER.info("usersWithCredentialsToUpdate size = {}", usersWithCredentialsToUpdate.size());
+        LOGGER.debug("usersWithCredentialsToUpdate = {}", usersWithCredentialsToUpdate);
```

Comment:
I don't think we need 2 log lines here. I mean I don't know if there is any benefit if we see the number of users to be updated during an investigation, but not the actual users. what do you think?

Reply:
I was following the pattern in this class, which emits two similar log statements (INFO with count, DEBUG with resource names) for all the various change sets it computes (users to add, users to delete, etc.). If we're always running with DEBUG logs enabled, then I guess there's not much point in logging the count. But otherwise, I think it makes sense to avoid bloating the production logs, and I do find it's useful to see the count, at least for debugging.

But I'm fine either way.